### PR TITLE
Updating corefx to use new roslyn compiler

### PR DESCRIPTION
**Pull request by joperezr**
cc: @weshaggard @ericstj 

FYI: @ahsonkhan @RussKeldorph @VSadov since you were interested in corefx using a newer compiler

This PR will update the version of the roslyn compiler that we use in corefx.

### CODE REVIEW

**Comment by JonHanna** [comment]
This reminds me that I meant to open an issue about discards not working in corefx. Will this fix that?

**Comment by joperezr** [comment]
It should since it will enable C#7 now.

**Comment by joperezr** [comment]
FYI: @stephentoub since this will make corefx be able to use C#7 features.

**Comment by ericstj** [comment]
Looks like desktop doesn't like the code generated for ImmutableCollections:
```
11:10:10       System.Collections.Immutable.Tests.ImmutableArrayBuilderTest.MoveToImmutableAfterClear [FAIL]
11:10:10         System.Security.VerificationException : Operation could destabilize the runtime.
11:10:10         Stack Trace:
11:10:10              at System.Collections.Immutable.ImmutableArray`1.Builder.ToArray()
11:10:10         System.Security.VerificationException : Operation could destabilize the runtime.
11:10:10         Stack Trace:
11:10:10              at System.Collections.Immutable.ImmutableHashSet`1.HashBucket.Enumerator.MoveNext()
11:10:10           D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Collections.Immutable\src\System\Collections\Immutable\ImmutableHashSet_1.Enumerator.cs(83,0): at System.Collections.Immutable.ImmutableHashSet`1.Enumerator.MoveNext()
11:10:10           D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Collections.Immutable\src\System\Collections\Immutable\ImmutableHashSet_1.cs(737,0): at System.Collections.Immutable.ImmutableHashSet`1.Overlaps(IEnumerable`1 other, MutationInput origin)
11:10:10           D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Collections.Immutable\src\System\Collections\Immutable\ImmutableHashSet_1.cs(372,0): at System.Collections.Immutable.ImmutableHashSet`1.Overlaps(IEnumerable`1 other)
11:10:10           D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Collections.Immutable\tests\ImmutableSetTest.cs(413,0): at System.Collections.Immutable.Tests.ImmutableSetTest.SetCompareTestHelper[T](Func`2 operation, Func`2 baselineOperation, IEnumerable`1 scenarios)
11:10:10         System.Security.VerificationException : Operation could destabilize the runtime.
11:10:10         Stack Trace:
11:10:10              at System.Collections.Immutable.ImmutableArray`1.Builder.ToArray()
11:10:10           D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Collections.Immutable\tests\ImmutableArrayBuilderTest.cs(443,0): at System.Collections.Immutable.Tests.ImmutableArrayBuilderTest.ToImmutable()
11:10:10         System.Security.VerificationException : Operation could destabilize the runtime.
11:10:10         Stack Trace:
11:10:10              at System.Collections.Immutable.ImmutableHashSet`1.HashBucket.Enumerator.MoveNext()
11:10:10           D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Collections.Immutable\src\System\Collections\Immutable\ImmutableHashSet_1.Enumerator.cs(83,0): at System.Collections.Immutable.ImmutableHashSet`1.Enumerator.MoveNext()
11:10:10              at System.Collections.Generic.HashSet`1.UnionWith(IEnumerable`1 other)
11:10:10              at System.Collections.Generic.HashSet`1..ctor(IEnumerable`1 collection, IEqualityComparer`1 comparer)
11:10:10           D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Collections.Immutable\tests\ImmutableSetTest.cs(546,0): at System.Collections.Immutable.Tests.ImmutableSetTest.ExceptTestHelper[T](IImmutableSet`1 set, T[] valuesToRemove)
```
Lots and lots of these.  Seems like it could be a compiler (or compiler+JIT) issue /cc @tmat

File: src/System.Runtime.Serialization.Formatters/tests/System.Runtime.Serialization.Formatters.Tests.csproj [hunk 1]
```diff
@@ -4,6 +4,7 @@
   <PropertyGroup>
     <ProjectGuid>{13CE5E71-D373-4EA6-B3CB-166FF089A42A}</ProjectGuid>
     <SkipIncludeNewtonsoftJson>true</SkipIncludeNewtonsoftJson>
+    <LangVersion>7.2</LangVersion>
```
Comment (ahsonkhan) [inline src/System.Runtime.Serialization.Formatters/tests/System.Runtime.Serialization.Formatters.Tests.csproj]: Quesiton: Would we need to update the csproj  of other assemblies (like System.Memory) to include LangVersion 7.2 if we want to use the ref keyword i.e. here: https://github.com/dotnet/corefx/pull/23908/files?

Comment (joperezr) [reply src/System.Runtime.Serialization.Formatters/tests/System.Runtime.Serialization.Formatters.Tests.csproj]: Yes, at least that was required for `ref structs` so for now I only did it in places where we used it. If we think we will want to use it everywhere, then we can set this property at the dir.props level later on so that it is not replicated on every project.

**Comment by stephentoub** [comment]
> Seems like it could be a compiler (or compiler+JIT) issue

cc: @VSadov

File: src/System.Runtime.Serialization.Formatters/tests/System.Runtime.Serialization.Formatters.Tests.csproj (hunk 1 above)
Comment (jkotas) [reply src/System.Runtime.Serialization.Formatters/tests/System.Runtime.Serialization.Formatters.Tests.csproj]: This should be added to the top level props file instead.

**Comment by VSadov** [comment]
I will take a look at ImmutableCollections. Interested what code hits that.

In some cases we may do optimizations that verifier does not like. My guess is that Immutable sets [SecurityTransparent] which causes verification at jit time and hits such optimization. Newer builds of compiler have an option to disable that. We talked with @joperezr and he will try newer build with a flag set for Immutable.

It could also be just a bug though :)

CC: @jaredpar

**Comment by jaredpar** [comment]
> Looks like desktop doesn't like the code generated for ImmutableCollection

what tool is being run here?

**Comment by stephentoub** [comment]
> what tool is being run here?

The error is coming from the runtime while executing the immutable collection tests in corefx.

**Comment by jkotas** [comment]
`System.Collections.Immutable` is marked as SecurityTransparent assembly (https://github.com/dotnet/corefx/pull/3847). SecurityTransparent assemblies have to be fully verifiable. In .NET Framework, the code is verified at runtime even when you are running in full trust. It is where the exception is coming from.

File: src/System.Runtime.Serialization.Formatters/tests/System.Runtime.Serialization.Formatters.Tests.csproj (hunk 1 above)
Comment (weshaggard) [reply src/System.Runtime.Serialization.Formatters/tests/System.Runtime.Serialization.Formatters.Tests.csproj]: Yes lets try and move that to dir.props.

File: src/Microsoft.XmlSerializer.Generator/tests/Microsoft.XmlSerializer.Generator.Tests.csproj [hunk 2]
```diff
@@ -54,7 +54,7 @@
       <SerializerName>$(AssemblyName).XmlSerializers</SerializerName>
     </PropertyGroup>
     <Message Text="Running Serialization Tool" Importance="normal" />
-    <Exec Command="$(GeneratorCliPath)dotnet $(OutputPath)dotnet-Microsoft.XmlSerializer.Generator.dll $(OutputPath)Microsoft.XmlSerializer.Generator.Tests.dll /force /casesensitive"  />
```
Comment (weshaggard) [inline src/Microsoft.XmlSerializer.Generator/tests/Microsoft.XmlSerializer.Generator.Tests.csproj]: Why is this change needed?

Comment (joperezr) [reply src/Microsoft.XmlSerializer.Generator/tests/Microsoft.XmlSerializer.Generator.Tests.csproj]: It is actually very weird, but I started getting an error saying that the /casesensitive parameter is not recognized

Comment (weshaggard) [reply src/Microsoft.XmlSerializer.Generator/tests/Microsoft.XmlSerializer.Generator.Tests.csproj]: OK.
cc @huanwu

**Comment by joperezr** [comment]
> Newer builds of compiler have an option to disable that. We talked with @joperezr and he will try newer build with a flag set for Immutable.

I have tried this with a newer version of the compiler and set the flag and now the tests pass just fine. I'll make the update on buildtools side and then update this PR to get all CI passing.

**Comment by joperezr** [comment]
@VSadov the netfx tests now pass just fine with the new compiler, but now one System.Reflection.Metadata test failed on all configurations with the following error:
```
     System.Reflection.Metadata.Decoding.Tests.SignatureDecoderTests.PinnedAndUnpinnedLocals [FAIL]
        Assert.Contains() Failure
        Not found: uint8& pinned
        In value:  ImmutableArray<String> ["uint8[]", "uint8*", "uint8[] pinned", "int32"]
```

Any idea what might be going on here? cc: @tmat

**Comment by jkotas** [comment]
@tmat is OOF.

It is fragile test that makes assumptions about the exact IL generated for `fixed`. It should be updated or related to reflect Roslyn changes.

**Comment by joperezr** [comment]
The failing test on that Linux machine is a random segfault that is not related to my PR. It is happening in System.Drawing on the Debian leg (cc: @mellinoe )
```
/home/helixbot/dotnetbuild/work/915f7222-9ffa-4f7f-8ce8-d6d5355597b1/Work/82dbbdcd-e7f1-4289-993a-4323d0b19f5c/Unzip/RunTests.sh: line 87: 116226 Segmentation fault      (core dumped) $RUNTIME_PATH/dotnet xunit.console.netcore.exe System.Drawing.Common.Tests.dll -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapptests -notrait category=nonlinuxtests -notrait category=OuterLoop -notrait category=failing
```

**Comment by joperezr** [comment]
Merging now since CI is now passing (other than that debian system.drawing test)

File: src/System.Runtime.Serialization.Formatters/tests/System.Runtime.Serialization.Formatters.Tests.csproj (hunk 1 above)
Comment (weshaggard) [reply src/System.Runtime.Serialization.Formatters/tests/System.Runtime.Serialization.Formatters.Tests.csproj]: @joperezr are you planning to move this to dir.props in a followup?

Comment (ahsonkhan) [reply src/System.Runtime.Serialization.Formatters/tests/System.Runtime.Serialization.Formatters.Tests.csproj]: If so, please also remove it from some of the places where I added it recently:
https://github.com/dotnet/corefx/pull/23908/files#diff-718e92a9d68dbfe993253e67ac1863f6R9
